tests/http_spec.rs

```diff
@@ -0,0 +1,215 @@
+        }
+      }
+      Ok(response)
+    }
+  }
+
+  #[tokio::test]
+  async fn test_body_yaml() {
+    let spec = HttpSpec::read("tests/data/sample.yaml").unwrap();
+    assert_downstream(spec).await;
+  }
+
+  async fn assert_downstream(spec: HttpSpec) {
+    for downstream_assertion in spec.clone().downstream_assertions.iter() {
+      if let Some(annotation) = downstream_assertion.annotation.clone() {
+        match annotation {
+          Annotation::Skip => {
+            let request_details = format!(
+              "Method: {:?}, Path: {}, Body: {}",
+              downstream_assertion.request.0.method.clone().unwrap_or_default(),
+              downstream_assertion
+                .request
+                .0
+                .url
+                .clone()
+                .unwrap_or(Url::parse("http://localhost:8080/graphql").unwrap()),
+              downstream_assertion
+                .request
+                .0
+                .body
+                .clone()
+                .unwrap_or(serde_json::Value::Null)
+            );
+            println!("Skipping test in: {}\nRequest Details: {}", spec.name, request_details);
+            continue;
+          }
+          Annotation::Only => {
+            let response = run(spec.clone(), &downstream_assertion).await.unwrap();
+            let body = hyper::body::to_bytes(response.into_body()).await.unwrap();
+            assert_eq!(
+              body,
+              serde_json::to_string(&downstream_assertion.response.0.body).unwrap()
+            );
+            break;
+          }
+          Annotation::Fail => {
+            let response = run(spec.clone(), &downstream_assertion).await;
+            assert!(response.is_err());
+          }
+        }
+      } else {
+        let response = run(spec.clone(), &downstream_assertion).await.unwrap();
+        let body = hyper::body::to_bytes(response.into_body()).await.unwrap();
+        assert_eq!(
+          body,
+          serde_json::to_string(&downstream_assertion.response.0.body).unwrap()
+        );
+      }
+    }
+  }
+
+  #[tokio::test]
+  async fn test_body_json() {
+    let spec = HttpSpec::read("tests/data/sample.json").unwrap();
+    assert_downstream(spec).await;
+  }
+
+  async fn run(spec: HttpSpec, downstream_assertion: &&DownstreamAssertion) -> anyhow::Result<hyper::Response<Body>> {
+    let query_string = serde_json::to_string(&downstream_assertion.request.0.body).unwrap();
+    let method = downstream_assertion.request.0.method.clone().unwrap_or_default();
+    let url = downstream_assertion
+      .request
+      .0
+      .url
+      .clone()
+      .unwrap_or(Url::parse("http://localhost:8080/graphql").unwrap());
+    let state = spec.setup().await;
+    let req = Request::builder()
+      .method(method)
+      .uri(url.as_str())
+      .body(Body::from(query_string))
+      .unwrap();
+    Ok(graphql_request(req, state.as_ref()).await?)
```

Comment:
sometimes instead of putting `unwrap` in many places it could be simper to use `?` and pass the error to the caller and on caller side we could have single unwrap or handle the error the way we want

Reply:
done